lib/src/git.rs

```diff
@@ -1305,53 +1317,110 @@ pub fn push_updates(
-            GitPushError::NoSuchRemote(remote_name.to_string())
-        } else {
-            GitPushError::InternalGitError(err)
-        }
-    })?;
-    let mut remaining_remote_refs: HashSet<_> = qualified_remote_refs.iter().copied().collect();
-    let mut push_options = git2::PushOptions::new();
-    let mut proxy_options = git2::ProxyOptions::new();
-    proxy_options.auto();
-    push_options.proxy_options(proxy_options);
-    let mut callbacks = callbacks.into_git();
-    callbacks.push_update_reference(|refname, status| {
-        // The status is Some if the ref update was rejected
-        if status.is_none() {
-            remaining_remote_refs.remove(refname);
+    let mut remaining_remote_refs_expected_locations =
+        qualified_remote_refs_expected_locations.clone();
+    let mut failed_push_negotiations = vec![];
+    let push_result = {
+        if remote_name == REMOTE_NAME_FOR_LOCAL_GIT_REPO {
+            return Err(GitPushError::RemoteReservedForLocalGitRepo);
         }
-        Ok(())
-    });
-    push_options.remote_callbacks(callbacks);
-    remote
-        .push(refspecs, Some(&mut push_options))
-        .map_err(|err| match (err.class(), err.code()) {
-            (git2::ErrorClass::Reference, git2::ErrorCode::NotFastForward) => {
-                GitPushError::NotFastForward
+        let mut remote = git_repo.find_remote(remote_name).map_err(|err| {
+            if is_remote_not_found_err(&err) {
+                GitPushError::NoSuchRemote(remote_name.to_string())
+            } else {
+                GitPushError::InternalGitError(err)
             }
-            _ => GitPushError::InternalGitError(err),
         })?;
-    drop(push_options);
-    if remaining_remote_refs.is_empty() {
-        Ok(())
-    } else {
-        Err(GitPushError::RefUpdateRejected(
-            remaining_remote_refs
-                .iter()
-                .sorted()
-                .map(|name| name.to_string())
-                .collect(),
+        let mut callbacks = callbacks.into_git();
+        let mut push_options = git2::PushOptions::new();
+        let mut proxy_options = git2::ProxyOptions::new();
+        proxy_options.auto();
+        push_options.proxy_options(proxy_options);
+        let failed_push_negotiations_ref = &mut failed_push_negotiations;
+        callbacks.push_negotiation(|updates| {
+            for update in updates {
+                // TODO: What would it mean if dst_refname was empty? AH, NEED TO USE dst
+                // INSTEAD!
+                let expected_location = update.dst_refname().and_then(|refname| {
+                    #[allow(suspicious_double_ref_op)]
+                    // TODO: Suggestions to remove the double clone are welcome.
+                    // First, Rust lulls you into a sense of security. Then, out
+                    // of nowhere, an hour is spent looking for an idiomatic way
+                    // of converting &&Option<CommitId> to Option<CommitId>.
```

Comment:
just `(**expr).clone()`?
I mean I guess it's also a bit ugly but that's what comes to mind when you ask for `&&Option<T> -> Option<T>`, and it's not worse than two clones

Reply:
Are you sure that works. IIRC, it did not.

Reply:
Since clone takes &self, it would autoref to `&**expr`, `*x` does not take ownership unconditionally (I had a similar intuition myself until encountering similar things fixed it for me).

I did try it locally before commenting to be sure, but just with rust-analyzer actually, let me `cargo build` it.. - seems to work ¯\\\_(ツ)_/¯

Reply:
Ah, I see.

I tried `**expr.get(...).unwrap().clone()`, and it didn't work. The error is:

```
error[E0614]: type `std::option::Option<backend::CommitId>` cannot be dereferenced
    --> lib/src/git.rs:1350:21
     |
1350 | /                     **qualified_remote_refs_expected_locations
1351 | |                         .get(refname)
1352 | |                         .expect("Push is trying to move a ref it wasn't asked to move")
1353 | |                         .clone()
     | |________________________________^
```

But `(**expr.get(...).unwrap()).clone()` works. Let me know if you know a way to understand this.

Reply:
Oh.. right :joy: 
Well `(*expr).clone()` is obvious and you understood it, we just turn &&T into &T and call a method that expects &T on it.

With `(**expr).clone()` I said it above, since we're calling a method that expects &T and we give it T it get's autoreferenced to &T, effectively turning our code into `(&**expr).clone()`, where `&*` cancel out and we get to the first option.
This "cancelling out" happens before ownership is considered I guess is the way to understand why it doesn't complain about that.

Rust auto(de)ref is joy, for example there's [this](https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md)

Reply:
Oh, I think I got it. `**expr.get(...).unwrap().clone()` means `**(expr.get(...).unwrap().clone())`. `expr.get(...).unwrap()` is of type `&&Option<CommitId>`. `that.clone()` is of type `&Option<CommitId>`. Dereferencing that once gives `Option<CommitId>`, and the second dereference causes that error.

If there wasn't a type error, the dereference still wouldn't work as rustc would complain that `Option<CommitId>` is not Copy. This is what you get on `*expr.get(...).unwrap().clone()`.

Reply:
> Rust auto(de)ref is joy, for example there's [this](https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md)

Yeah, that's why I spent an hour on this. I kept trying various combinations of `.as_ref()`, `.as_deref()`, 🤯

Reply:
> what's going on in the former case

oh lol I thought you needed me to expand on the `(**expr).clone()` thing, well you figured it out ¯\\\_(ツ)_/¯